File: agenkit-cpp/examples/observability_basic.cpp
/**
 * @file observability_basic.cpp
 * @brief Basic observability example with tracing, metrics, logging, and audit
 *
 * This example demonstrates:
 * - Initializing OpenTelemetry tracing and metrics
 * - Configuring structured logging
 * - Setting up audit logging
 * - Wrapping agents with observability middleware
 * - Processing messages with full observability
 */

#ifdef AGENKIT_WITH_OBSERVABILITY

#include "agenkit/observability/tracing.hpp"
#include "agenkit/observability/metrics.hpp"
#include "agenkit/observability/logging.hpp"
#include "agenkit/observability/audit.hpp"
#include "agenkit/adapters/echo_agent.hpp"
#include <iostream>
#include <memory>

using namespace agenkit;
using namespace agenkit::observability;

int main() {
    std::cout << "=== Agenkit C++ Observability - Basic Example ===" << std::endl;

    // Step 1: Initialize tracing (console exporter for demo)
    std::cout << "\n1. Initializing tracing..." << std::endl;
    try {
        init_tracing("console", "");
        std::cout << "   ✓ Tracing initialized (console exporter)" << std::endl;
    } catch (const std::exception& e) {
        std::cerr << "   ✗ Failed to initialize tracing: " << e.what() << std::endl;
        return 1;
    }

    // Step 2: Initialize metrics (console exporter for demo)
    std::cout << "\n2. Initializing metrics..." << std::endl;
    try {
        init_metrics("console", "");
        std::cout << "   ✓ Metrics initialized (console exporter)" << std::endl;
    } catch (const std::exception& e) {
        std::cerr << "   ✗ Failed to initialize metrics: " << e.what() << std::endl;
        return 1;
    }

    // Step 3: Configure structured logging
    std::cout << "\n3. Configuring logging..." << std::endl;
    try {
        configure_logging("pretty", "info");
        std::cout << "   ✓ Logging configured (pretty format, info level)" << std::endl;
    } catch (const std::exception& e) {
        std::cerr << "   ✗ Failed to configure logging: " << e.what() << std::endl;
        return 1;
    }

    // Step 4: Create audit logger
    std::cout << "\n4. Creating audit logger..." << std::endl;
    std::shared_ptr<AuditLogger> audit;
    try {
        audit = AuditLogger::create("audit_basic.log", 10);
        std::cout << "   ✓ Audit logger created (file: audit_basic.log)" << std::endl;
    } catch (const std::exception& e) {
        std::cerr << "   ✗ Failed to create audit logger: " << e.what() << std::endl;
        return 1;
    }

    // Step 5: Create and wrap agent with observability
    std::cout << "\n5. Creating observable agent..." << std::endl;
    auto echo = std::make_shared<EchoAgent>();
    std::cout << "   ✓ Created EchoAgent" << std::endl;

    auto traced = std::make_shared<TracingMiddleware>(echo, "echo.process");
    std::cout << "   ✓ Wrapped with TracingMiddleware" << std::endl;

    auto observed = std::make_shared<MetricsMiddleware>(traced);
    std::cout << "   ✓ Wrapped with MetricsMiddleware" << std::endl;

    // Audit agent creation
    audit->log(
        AuditEvent::create(AuditEventType::AgentCreated, "echo_agent", "demo_session")
            .with_detail("middleware", "tracing,metrics")
            .with_detail("example", "basic")
            .with_severity(Severity::INFO)
    );
    std::cout << "   ✓ Audited agent creation" << std::endl;

    // Step 6: Process messages
    std::cout << "\n6. Processing messages..." << std::endl;

    std::vector<std::string> messages = {
        "Hello, observability!",
        "This message is traced",
        "Metrics are being recorded"
    };

    for (size_t i = 0; i < messages.size(); i++) {
        std::cout << "\n   Message " << (i + 1) << ": \"" << messages[i] << "\"" << std::endl;

        // Log the event
        log_agent_event("message_received", "Processing message " + std::to_string(i + 1));

        // Create message
        Message msg;
        msg.role = "user";
        msg.content = messages[i];

        // Process (creates span, records metrics)
        auto result_future = observed->process(msg);
        auto result = result_future.get();

        if (result.is_ok()) {
            auto response = result.unwrap();
            std::cout << "   ✓ Response: \"" << response.content << "\"" << std::endl;

            // Log success
            log_agent_event("message_processed", "Successfully processed message " + std::to_string(i + 1));

            // Audit the operation
            audit->log(
                AuditEvent::create(AuditEventType::MessageProcessed, "echo_agent", "demo_session")
                    .with_detail("message_index", std::to_string(i + 1))
                    .with_detail("message_length", std::to_string(messages[i].length()))
                    .with_severity(Severity::INFO)
            );
        } else {
            std::cerr << "   ✗ Error: " << result.error().message() << std::endl;

            // Log error
            log_agent_error("message_failed", "Failed to process message " + std::to_string(i + 1),
                          result.error().message());

            // Audit the failure
            audit->log(
                AuditEvent::create(AuditEventType::MessageFailed, "echo_agent", "demo_session")
                    .with_detail("error", result.error().message())
                    .with_severity(Severity::ERROR)
            );
        }
    }

    // Step 7: Flush audit log
    std::cout << "\n7. Flushing audit log..." << std::endl;
    audit->flush();
    std::cout << "   ✓ Audit log flushed to disk" << std::endl;

    // Step 8: Query audit events
    std::cout << "\n8. Querying audit events..." << std::endl;
    auto events = audit->query();
    std::cout << "   ✓ Total events: " << events.size() << std::endl;

    auto created_events = audit->query_by_type(AuditEventType::AgentCreated);
    std::cout << "   ✓ Agent creation events: " << created_events.size() << std::endl;

    auto processed_events = audit->query_by_type(AuditEventType::MessageProcessed);
    std::cout << "   ✓ Message processed events: " << processed_events.size() << std::endl;

    // Step 9: Summary
    std::cout << "\n=== Summary ===" << std::endl;
    std::cout << "✓ Tracing: Spans created for each message" << std::endl;
    std::cout << "✓ Metrics: Request count and duration recorded" << std::endl;
    std::cout << "✓ Logging: Events logged with trace context" << std::endl;
    std::cout << "✓ Audit: " << events.size() << " events persisted to audit_basic.log" << std::endl;
    std::cout << "\nCheck audit_basic.log for detailed audit trail!" << std::endl;

    return 0;
}

#else

#include <iostream>

int main() {
    std::cerr << "Error: This example requires AGENKIT_WITH_OBSERVABILITY=ON" << std::endl;
    std::cerr << "Rebuild with: cmake -DAGENKIT_WITH_OBSERVABILITY=ON .." << std::endl;
    return 1;
}

#endif // AGENKIT_WITH_OBSERVABILITY

File: agenkit-rust/examples/observability_basic.rs
//! Basic observability example with console exporters.
//!
//! This example demonstrates:
//! - Setting up tracing with console exporter
//! - Setting up metrics with stdout exporter
//! - Configuring structured logging
//! - Using TracingMiddleware and MetricsMiddleware
//! - Basic audit logging
//!
//! Run with:
//! ```bash
//! cargo run --example observability_basic
//! ```

use agenkit::core::{Agent, AgentError, Message};
use agenkit::observability::{
    configure_logging, init_metrics, init_tracing, log_agent_event, AuditEvent, AuditEventType,
    AuditLogger, MetricsMiddleware, TracingMiddleware,
};
use async_trait::async_trait;
use std::collections::HashMap;
use tempfile::TempDir;

/// Simple echo agent for demonstration
struct EchoAgent {
    name: String,
}

impl EchoAgent {
    fn new(name: &str) -> Self {
        Self {
            name: name.to_string(),
        }
    }
}

#[async_trait]
impl Agent for EchoAgent {
    fn name(&self) -> &str {
        &self.name
    }

    async fn process(&self, message: Message) -> Result<Message, AgentError> {
        // Echo back the message with a prefix
        let response = format!("Echo: {}", message.content);
        Ok(Message::with_text("assistant", &response))
    }
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    println!("=== Agenkit Rust - Basic Observability Example ===\n");

    // Step 1: Initialize tracing with console exporter
    println!("1. Initializing distributed tracing...");
    init_tracing("console", None)?;
    println!("   ✓ Tracing initialized (console exporter)\n");

    // Step 2: Initialize metrics with stdout exporter
    println!("2. Initializing metrics...");
    init_metrics("stdout", None)?;
    println!("   ✓ Metrics initialized (stdout exporter)\n");

    // Step 3: Configure structured logging
    println!("3. Configuring structured logging...");
    configure_logging("pretty", "info")?;
    println!("   ✓ Logging configured (pretty format, info level)\n");

    // Step 4: Set up audit logging
    println!("4. Setting up audit logging...");
    let temp_dir = TempDir::new()?;
    let audit_path = temp_dir.path().join("audit.log");
    let audit_logger = AuditLogger::new(audit_path.clone());
    println!("   ✓ Audit logger created: {:?}\n", audit_path);

    // Step 5: Create agent with observability middleware
    println!("5. Creating instrumented agent...");
    let agent = EchoAgent::new("echo-agent");
    let traced_agent = TracingMiddleware::new(agent, None);
    let observed_agent = MetricsMiddleware::new(traced_agent);
    println!("   ✓ Agent wrapped with TracingMiddleware and MetricsMiddleware\n");

    // Step 6: Log agent creation
    let audit_event = AuditEvent::new(
        AuditEventType::AgentCreated,
        "echo-agent".to_string(),
        Some("demo-session".to_string()),
    );
    audit_logger.log(audit_event).await?;

    // Step 7: Process messages
    println!("6. Processing messages with full observability...\n");

    let messages = vec!["Hello, World!", "How are you?", "Testing observability"];

    for (i, content) in messages.iter().enumerate() {
        println!("   Message {}: \"{}\"", i + 1, content);

        // Log the event
        let mut details = HashMap::new();
        details.insert("message_id".to_string(), serde_json::json!(i + 1));
        details.insert("content".to_string(), serde_json::json!(content));
        log_agent_event("message.processing", &details);

        // Process the message (tracing span created automatically)
        let message = Message::with_text("user", *content);
        let response = observed_agent.process(message).await?;

        println!("   Response: \"{}\"", response.content_as_str().unwrap_or(""));

        // Log audit event
        let audit_event = AuditEvent::new(
            AuditEventType::MessageProcessed,
            "echo-agent".to_string(),
            Some("demo-session".to_string()),
        );
        audit_logger.log(audit_event).await?;

        println!();
    }

    // Step 8: Flush audit logs
    println!("7. Flushing audit logs...");
    audit_logger.flush().await?;
    println!("   ✓ Audit logs written to disk\n");

    // Step 9: Query audit events
    println!("8. Querying audit events...");
    let events = audit_logger
        .query(Some("demo-session".to_string()))
        .await?;
    println!("   ✓ Found {} audit events for session 'demo-session'", events.len());

    for (i, event) in events.iter().enumerate() {
        println!(
            "     Event {}: {:?} - {} at {}",
            i + 1,
            event.event_type,
            event.agent_name,
            event.timestamp.format("%Y-%m-%d %H:%M:%S")
        );
    }

    println!("\n=== Example Complete ===");
    println!("This example demonstrated:");
    println!("  • Distributed tracing with OpenTelemetry");
    println!("  • Metrics collection (request counts, durations)");
    println!("  • Structured logging with trace correlation");
    println!("  • Audit logging with session tracking");
    println!("\nAll observability data was captured during agent execution!");

    Ok(())
}
